src/network/networkd-routing-policy-rule.c

```diff
@@ -0,0 +1,904 @@
+/***
+  This file is part of systemd.
+
+  Copyright 2017 Susant Sahani
+
+  systemd is free software; you can redistribute it and/or modify it
+  under the terms of the GNU Lesser General Public License as published by
+  the Free Software Foundation; either version 2.1 of the License, or
+  (at your option) any later version.
+
+  systemd is distributed in the hope that it will be useful, but
+  WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public License
+  along with systemd; If not, see <http://www.gnu.org/licenses/>.
+***/
+
+#include <net/if.h>
+#include <linux/fib_rules.h>
+
+#include "alloc-util.h"
+#include "conf-parser.h"
+#include "fileio.h"
+#include "networkd-routing-policy-rule.h"
+#include "netlink-util.h"
+#include "networkd-manager.h"
+#include "parse-util.h"
+#include "socket-util.h"
+#include "string-util.h"
+
+int routing_policy_rule_new(RoutingPolicyRule **ret) {
+        RoutingPolicyRule *rule;
+
+        rule = new0(RoutingPolicyRule, 1);
+        if (!rule)
+                return -ENOMEM;
+
+        rule->family = AF_INET;
+        rule->table = RT_TABLE_MAIN;
+
+        *ret = rule;
+        rule = NULL;
+
+        return 0;
+}
+
+void routing_policy_rule_free(RoutingPolicyRule *rule) {
+
+        if (!rule)
+                return;
+
+        if (rule->network) {
+                LIST_REMOVE(rules, rule->network->rules, rule);
+                assert(rule->network->n_rules > 0);
+                rule->network->n_rules--;
+
+                if (rule->section) {
+                        hashmap_remove(rule->network->rules_by_section, rule->section);
+                        network_config_section_free(rule->section);
+                }
+
+                if (rule->network->manager) {
+                        set_remove(rule->network->manager->rules, rule);
+                        set_remove(rule->network->manager->rules_foreign, rule);
+                }
+        }
+
+        free(rule);
+}
+
+static void routing_policy_rule_hash_func(const void *b, struct siphash *state) {
+        const RoutingPolicyRule *rule = b;
+
+        assert(rule);
+
+        siphash24_compress(&rule->family, sizeof(rule->family), state);
+
+        switch (rule->family) {
+        case AF_INET:
+        case AF_INET6:
+
+                if (!in_addr_is_null(rule->family, &rule->from)) {
```

Comment:
Do we actually need the ifs? Not that it matters much, but I'd guess that the conditional costs more than the hashing of a few bytes, and the code would be slightly shorter.